consensus/core/src/leader_schedule.rs

```diff
@@ -63,28 +83,208 @@ impl LeaderSchedule {
 
         leader_index
     }
+
+    /// Atomically updates the `LeaderSwapTable` with the new provided one. Any
+    /// leader queried from now on will get calculated according to this swap
+    /// table until a new one is provided again.
+    fn update_leader_swap_table(&self, table: LeaderSwapTable) {
+        tracing::trace!("Updating {:?}", table);
+
+        let mut write = self.leader_swap_table.write();
+        *write = table;
+    }
+}
+
+#[derive(Default, Clone)]
+pub(crate) struct LeaderSwapTable {
```

Comment:
Should LeaderSwapTable also keep track of the commit range where it is computed from? Then we can add some validations to log and error when the swap table is used for unintended rounds.